sdks/python/apache_beam/transforms/trigger.py

```diff
@@ -1180,6 +1181,11 @@ def merge(_, to_be_merged, merge_result):  # pylint: disable=no-self-argument
         for window in merged_away:
           state.clear_state(window, self.WATERMARK_HOLD)
 
+    # drop expired windows
+    for window in list(windows_to_elements.keys()):
+      if input_watermark > window.end + self.allowed_lateness:
+        del windows_to_elements[window]
```

Comment:
This test is not on the element timestamp, but window end. It seems it would be cleaner to simply not add it up at line 1151. (We need it up there to avoid merging on "unseen" data as well).

Reply:
Just to double check: for a WindowValue that is expiring before merging but not expiring after merging, is the expected behavior to garbage collect it anyways?

Reply:
Yes. Once it's fallen behind, it can be dropped at any point.